File: programmers/kakao#2/python/KAKAO30.py
import heapq

def solution(n, k, cmds):
    left, right, removed = [-i for i in range(k + 1)], [i for i in range(k+1, n)], []
    
    heapq.heapify(left)
    heapq.heapify(right)
    
    for cmd in cmds:
        if cmd[0] == "U":
            cnt = int(cmd.split(" ")[1])
            
            for _ in range(cnt):
                heapq.heappush(right, -heapq.heappop(left))
                
        elif cmd[0] == "D":
            cnt = int(cmd.split(" ")[1])
            
            for _ in range(cnt):
                heapq.heappush(left, -heapq.heappop(right))
                
        elif cmd[0] == "C":
            removed.append(-heapq.heappop(left))
            
            if right:
                heapq.heappush(left, -heapq.heappop(right))
        else:
            num = removed.pop()
            if num > -left[0]:
                heapq.heappush(right, num)
            else:
                heapq.heappush(left, -num)
                
    answer = ["O" for _ in range(n)]
    while removed:
        answer[removed.pop()] = "X"
    return "".join(answer)

'''
[카카오 2021 인턴] 표 편집
해결: O
'''

File: 0_4.programmers/kakao#2/java/KAKAO30.java
import java.util.*;
class KAKAO30 {
    public String solution(int n, int k, String[] cmds) {
        PriorityQueue<Integer> left = new PriorityQueue<>((o1, o2) -> o2 - o1);
        PriorityQueue<Integer> right = new PriorityQueue<>((o1, o2) -> o1 - o2);
        Stack<Integer> removed = new Stack<>();
        
        // 처음에 선택된 행
        for (int i=0;i<=k;i++) {
            left.add(i);
        }
        
        for (int i=k+1;i<n;i++) {
            right.add(i);
        }
        
        
        for (String cmd : cmds) {
            String c = cmd.split(" ")[0];
            
            if (c.equals("U")) {
                int num = Integer.parseInt(cmd.split(" ")[1]);
                
                for (int i=0;i<num;i++) {
                    right.add(left.poll());
                }
            } else if (c.equals("D")) {
                int num = Integer.parseInt(cmd.split(" ")[1]);
                
                for (int i=0;i<num;i++) {
                    left.add(right.poll());
                }
            } else if (c.equals("C")) {
                removed.add(left.poll());
                
                if (!right.isEmpty())
                    left.add(right.poll());
                
            } else {
                if (removed.peek() > left.peek()) {
                    right.add(removed.pop());
                } else {
                    left.add(removed.pop());
                }
            }
        }
        
        boolean[] removedMark = new boolean[n];
        for (int i=0;i<removed.size();i++) {
            removedMark[removed.get(i)] = true;
        }
        
        StringBuilder sb = new StringBuilder();
        
        for (int i=0;i<n;i++) {
            if (!removedMark[i])
                sb.append("O");
            else
                sb.append("X");
        }
        return sb.toString();
    }
}
/**
 * [카카오 2021 인턴] 표 편집
 * 해결: O
 */